### PR TITLE
Skip PVC restore if already existing

### LINKED ISSUE #340: Restore of a namespace with existing PVCs is going in infinite loop
velero-plugin-for-vsphere version 1.1.0
Velero version 1.5.4

**Steps performed:**
1. Created a new Namespace with PVC. (Used the sample provided in Velero package - nginx with PV).
2. Performed backup of this Namespace.
3. Did not make any changes to the Namespace and triggered a restore operation using the above backup with all the resources included.
4. Restore operation gets stuck in InProgress state.

**Expected behaviour:** Restore operation should be successful.

**Few pointers which can help with the resolution of this issue:**
1. The original PVCs were not changed and they still exist with the same name on the cluster.
2. the same operation was working fine with velero-plugin-for-vsphere version 1.0.2 and Velero version 1.5.4. The restore operation completes and provides warnings for skipping resources due to conflicts.
3. Download snapshot operations seems to be failing with the below error:
    "Failed to download snapshot, ivd:0d79bc4c-2a06-4be1-83f4-49c2e087c858:4e8fbdb2-8ace-4f25-a119-57de0cbf5553, from durable object storage. Data copy failed from PE ivd:0d79bc4c-2a06-4be1-83f4-49c2e087c858, to PE ivd:0d79bc4c-2a06-4be1-83f4-49c2e087c858:4e8fbdb2-8ace-4f25-a119-57de0cbf5553: Open virtual disk file failed. The error code is 1. with error code: 1"
4. After the above failure download operation seems to be getting retried few more times. (6 times) Post this it gets marked as failed.
5. After every 20-21 minutes a new download CR is getting generated and the download operations gets triggered again with a new CR.
6. Restore operation never completes. Monitored for over 2 days. 
7. If I delete the Velero restore job and fire a new restore job (with the same backup or different), then this job gets stuck in the New phase. This may be because earlier CloneFromSnapshot is still in the pending state.

Logs files attached with debug logs enabled.
[RestoreNapespaceIssueLogs.zip](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/6347866/RestoreNapespaceIssueLogs.zip)

### PR DESCRIPTION
**Pull request by xing-yang**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #340

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Skips PVC restore if it already exists.
```

### CODE REVIEW

**Comment by xing-yang** [comment]
# Testing
## Normal restore succeeded
```
velero restore describe test-restore

Name:         test-restore
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  Completed

Started:    2021-06-03 08:00:23 +0000 UTC
Completed:  2021-06-03 08:02:12 +0000 UTC

Backup:  test-backup

Namespaces:`
  Included:  all namespaces found in the backup
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```

## Restore right after backup without deleting any resources
```
velero restore describe new-restore

Name:         new-restore
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  Completed

Started:    2021-06-03 13:53:50 +0000 UTC
Completed:  2021-06-03 13:53:52 +0000 UTC

Warnings:
  Velero:     <none>
  Cluster:    <none>
  Namespaces:
    test-ns-gleeojz:  could not restore, persistentvolumeclaims "etcd0-pv-claim" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, persistentvolumeclaims "kibishii-data-kibishii-deployment-0" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, persistentvolumeclaims "kibishii-data-kibishii-deployment-1" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, pods "etcd0" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, pods "jump-pad" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, pods "kibishii-deployment-0" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, pods "kibishii-deployment-1" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, services "etcd-client" already exists. Warning: the in-cluster version is different than the backed-up version.
                      could not restore, services "etcd0" already exists. Warning: the in-cluster version is different than the backed-up version.

Backup:  new-backup

Namespaces:
  Included:  all namespaces found in the backup
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```

File: pkg/plugin/restore_pvc_action_plugin.go [hunk 1]
```diff
@@ -140,6 +142,31 @@ func (p *NewPVCRestoreItemAction) Execute(input *velero.RestoreItemActionExecute
 		return nil, errors.WithStack(err)
 	}
 
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
```
Comment (lintongj) [inline pkg/plugin/restore_pvc_action_plugin.go:145]: Would you please create a util function under pkg/plugin/util/ for the code snippet below? The `RetrieveStorageClassMapping` func below would be an example.

File: pkg/plugin/restore_pvc_action_plugin.go [hunk 2]
```diff
@@ -140,6 +142,31 @@ func (p *NewPVCRestoreItemAction) Execute(input *velero.RestoreItemActionExecute
...
+		p.Log.Errorf("Error occurred when trying to get PVC %s/%s: %v", pvc.Namespace, pvc.Name, err)
+		return nil, errors.WithStack(err)
+	} else if getPvc != nil && getPvc.Namespace == pvc.Namespace && getPvc.Name == pvc.Name {
+		// NOTE: If PVC does not exist, "Get" may return an empty PersistentVolumeClaim object, not nil; so we need to check Namespace/Name match here
+		// NOTE: Need to skip it here in the plugin to avoid hanging as Velero skips restoring already existing resources
+		p.Log.Warnf("Skipping PVCRestoreItemAction for PVC %s/%s since it already exists.", pvc.Namespace, pvc.Name)
+		return &velero.RestoreItemActionExecuteOutput{
+			UpdatedItem: input.Item,
+		}, nil
+	} else if err != nil && apierrs.IsAlreadyExists(err) {
```
Comment (lintongj) [inline pkg/plugin/restore_pvc_action_plugin.go:161]: What is the case where a GET API returns `IsAlreadyExists` error? I wonder if it is possible.

Comment (xing-yang) [reply pkg/plugin/restore_pvc_action_plugin.go]: Removed this.  I think this happens when creating a resource.

File: pkg/plugin/restore_pvc_action_plugin.go (hunk 1 above)
Comment (xing-yang) [reply pkg/plugin/restore_pvc_action_plugin.go]: Done

**Comment by xing-yang** [comment]
@lintongj Comments are addressed.  PTAL.

**Comment by svcbot-qecnsdp** [comment]
Congratulations !! 

All the pre-check tests passed
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/549/

**Review by lintongj — APPROVED** [review]

**Pull request merged**